[android] Status bar improvements

arnaudvergnet opened the pull request:
This PR does 2 things:
- First it makes the status bar opaque for every activity/fragment except the map activity to avoid strange workarounds to mimic an opaque status bar. For the launch screen, the status bar is now the same color as the background for consistency with the navigation bar.  For the initial download screen, the status bar is now black.
- Then it changes the way the status bar is handled by using the insets instead of a resource integer. The status bar color is left to the system so it is always possible to read it and its color is consistent with the navigation bar. This should prove more robust and fix bugs on some phones.

~~Sadly point 2 introduced a bug I am still trying to fix. The bottom sheet menus on the map (layers and main menu) show a weird animation with the status bar (see 00:26 in the video). This is not happening in activities with the status bar opaque. I'll keep this PR in draft until I can fix this but in the meantime help in testing would be much appreciated to make sure all the issues linked are indeed fixed.~~
**EDIT: this is now fixed**

I also removed a Java class `EditDescriptionFragment.java` which I missed in the last cleanup MR. It was referring to status bar utilities I removed so I removed the class as well.

@rtsisyk can you confirm it fixes https://github.com/organicmaps/organicmaps/issues/3411?

@najodleglejszy would you be available to test a debug apk and confirm it fixes your issue?


https://user-images.githubusercontent.com/80701113/192308841-476f6a34-cdb1-4004-8bd0-dea547c81c42.mp4

![Screenshot_1664269627](https://user-images.githubusercontent.com/80701113/192485991-12c3304b-4ab6-48e9-a8a2-0b8f704895dc.png)


Fixes https://github.com/organicmaps/organicmaps/issues/3411
Fixes https://github.com/organicmaps/organicmaps/issues/3444
Fixes https://github.com/organicmaps/organicmaps/issues/1654
Fixes https://github.com/organicmaps/organicmaps/issues/3393
Fixes https://github.com/organicmaps/organicmaps/issues/3611

arnaudvergnet commented:
Thanks! No worries I wanted to see if you wanted to test before uploading the apk. Here it is: https://github.com/arnaudvergnet/organicmaps-apk-test/raw/main/OrganicMaps-22092602-fdroid-debug-PR-3495.apk.

@biodranik can you confirm the app installed with a debug apk will not temper with the release version?

najodleglejszy commented:
yeah, that looks better! I've noticed the status bar is darker and/or more opaque than compared to the current version, but the height is definitely the way it should be now.

![Screenshot_20220926-181358](https://user-images.githubusercontent.com/8598559/192328285-c2d6252b-2a91-4fa5-8d97-40e254914b09.png)

arnaudvergnet commented:
Yes this is darker on purpose because in some situations the text inside it would be too difficult to read. Nice to see it fixed your issue thanks for testing!

arnaudvergnet commented:
PR is now ready for review I fixed all the bugs I could find. Please tell me if you find any issue about the status bar which I missed in the main post.

biodranik commented on android/src/com/mapswithme/util/UiUtils.java:
```diff
@@ -363,36 +351,21 @@ public static boolean isViewTouched(@NonNull MotionEvent event, @NonNull View vi
-    if (lp.height == ViewGroup.LayoutParams.WRAP_CONTENT)
-    {
+    view.setOnApplyWindowInsetsListener((v, windowInsets) -> {
+      int statusBarHeight = windowInsets.getSystemWindowInsetTop();
```
final

biodranik commented on android/src/com/mapswithme/maps/widget/placepage/EditDescriptionFragment.java:
```diff
@@ -1,125 +0,0 @@
-    {
-      final String descriptionNoSimpleTags = StringUtils.removeEditTextHtmlTags(description);
-      if (!StringUtils.nativeIsHtml(descriptionNoSimpleTags))
-        description = Html.fromHtml(description).toString();
```
Wait, is the bookmark's description now editable?

arnaudvergnet replied:
Yes I really don't know what this class was used for

biodranik reviewed (changes requested):
1. Does bookmark description editing works properly for HTML and non-HTML descriptions?
2. Does it work with our https://omaps.app/api ?
3. @rtsisyk can you please test these changes?

arnaudvergnet commented:
Editing a bookmark's description

https://user-images.githubusercontent.com/80701113/192737028-2bbc1a9b-8260-4ddc-a2e4-dc5a0fe8fe0d.mp4

Using the Omaps API

https://user-images.githubusercontent.com/80701113/192737072-0de6e45d-9d22-4d3a-926a-1d95ca91d1fb.mp4

biodranik commented:
@arnaudvergnet 

1. Will it simplify the code if the map's status bar will be transparent, as on iOS?

2. Is it hard to change the background color of the status bar to the one used in OM's dialog, like here?

![image](https://user-images.githubusercontent.com/170263/192850593-970103c8-891c-42de-9af6-e5875a142b26.jpeg)

arnaudvergnet commented:
> Will it simplify the code if the map's status bar will be transparent, as on iOS?

I don't think it would change much. But I really think this is a bad idea, with the background in current master it is already hard enough to read, if you remove it completely it would be worse.

> Is it hard to change the background color of the status bar to the one used in OM's dialog, like here?

So you don't want any color difference between the header and the status bar right? This is easy to do on all activities except the map one. I'll have to think a bit about how to proceed for the map activity. In this PR the status bar is set to translucent and I cannot control the color.

I would need to make it fully transparent when the picker or the routing toolbars are shown. So to support this case making the toolbar fully transparent would make things easier but as I said, I do not think it would be a good choice.

biodranik commented:
I understand your worries about the visibility. But the status bar is not the main thing in the maps app. It should not distract users from their primary goal: to find something on the map. On iOS status bar texts are not always visible, but that's not a big deal, because you can easily drag the map and immediately see it. And as I mentioned, it is used there already for many years. Nobody complained so far )

It would be great to try the transparent status bar on an actual device. Then it would be easier to approve this change.

arnaudvergnet commented:
I tried a bit the transparent toolbar, and it looks good if the text is black. Sadly this is only available for api > 23 (we support up to 21). So I will have to do a little workaround on api 21-22 for the status bar to be visible on the white background of the navigation header.

biodranik commented:
The simplest solution would be enough. It is ok if older devices will look different if the code will be simpler. We don't have too much such devices, and their number will decrease over time.

arnaudvergnet commented:
I will at least try to make the status bar readable in navigation for those old devices because this can be really annoying. Hopefully it will not make the code too complex.

biodranik commented:
Did you check [this post](https://stackoverflow.com/questions/49254877/white-navigationbar-with-grey-icons/49255140#49255140)?

arnaudvergnet commented:
> Did you check [this post](https://stackoverflow.com/questions/49254877/white-navigationbar-with-grey-icons/49255140#49255140)?

Yes but I was not able to make this work. I'll have to dig a bit deeper.

arnaudvergnet commented:
I managed to do something. On API < 23 I added a translucent background to the status bar when its content would be black on API >= 23. Here is how it looks on API 21 and 30:

https://user-images.githubusercontent.com/80701113/193254220-be4082f2-42af-4be9-ae14-8eca31e9874b.mp4

https://user-images.githubusercontent.com/80701113/193254554-646eebfb-0d6f-4796-8fd0-e0585dc62e08.mp4

There is so much difference on how to handle system insets since API 23 that I had to create "legacy" functions to handle pre API 23 cases. I also noticed some bugs on those old devices were introduced by https://github.com/organicmaps/organicmaps/pull/3368. I fixed some but as you may see in the first video, there are still some problems with the navigation menu.

I'll put this PR as draft while I fix those and clean up the code. But in the meantime can you give me some feedback on those results?

arnaudvergnet commented:
@biodranik I fixed every issue with this PR I could find. I'll remove the draft status and reorder the commits when you are satisfied with the changes.

biodranik commented:
> I'll put this PR as draft while I fix those and clean up the code. But in the meantime can you give me some feedback on those results?

Looks great!

biodranik commented:
Tested on Android 8, it looks very cool, thanks!

Two issues:
1. Black text is not visible on the map in the dark mode (but it changes to a light one on other screens). It should be lighter, probably not too bright, either in all dialogs or at least on the map (to avoid distraction when driving in a car at night).
2. Layers animation immediately switches dark text to the light one, and it's great. But opening other animated dialogs switches dark text to light one only at the end of the animation. Is it possible to do it at the start of the opening animation? (not sure about the closing animation, it looks more or less ok, maybe it is possible to compare it and see if switching colors at the start of the closing animation is better).

arnaudvergnet commented:
For 1 yes I just noticed I forgot to support dark mode. Please note we cannot control finely the color of the text in the status bar: it is either black or white.

For 2, I am not sure I understand I'll have to try and see on a device. As we use the standard android components for the bottom sheets, they also use standard animations. So I would keep them as it is and avoid customizing too much to prevent the code from being too complex.

biodranik commented:
RE:2 Bottom sheets do not matter. The current behavior is:
1. Press a button that opens bookmarks/search/settings.
2. Animation starts.
3. Animation ends.
4. Toolbar color changes.

I think that this one will look better:
1. Press a button that opens bookmarks/search/settings.
2. Toolbar color changes.
3. Animation starts.
4. Animation ends.

arnaudvergnet commented:
This is not really something we can control easily because it is tied to the system's animations, which change depending on the Android/OEM version. On Android 11  for example, I cannot reproduce your issue. The status bar color changes automatically depending on the activity' style. Manually changing it before the new activity is started would prove difficult and prone to introducing new bugs.

https://user-images.githubusercontent.com/80701113/193403433-84375b85-f573-4c1a-9ef7-1551aa44c61a.mp4

biodranik commented:
Ok, then let's leave it for the future. On Android 8 dialogs are opened slower, and from right to left. Maybe it's possible to make their appearance animation faster?

arnaudvergnet commented:
I don't know from the app level, unless we create our own animations (which I do not recommend). It is a setting you can control globally from the developer options.

arnaudvergnet commented:
Here is how the night mode looks like:
|||||
|-|-|-|-|
| ![Screenshot_1664632354](https://user-images.githubusercontent.com/80701113/193412861-fefb5a02-4b0c-467d-95b5-53b03cc5f1c8.png) | ![Screenshot_1664632364](https://user-images.githubusercontent.com/80701113/193412863-049b63a4-74a6-4b6a-81dc-bb18e645e2be.png) | ![Screenshot_1664632368](https://user-images.githubusercontent.com/80701113/193412864-23422880-b754-42c9-9fb2-dccebbc70e16.png) | ![Screenshot_1664632380](https://user-images.githubusercontent.com/80701113/193412866-1a2f4df5-a952-43d2-9453-91ad31715f2d.png) |

arnaudvergnet commented:
@biodranik the git history is better now. Before merging I would love to get some help for testing to make sure no new issues are introduced. System insets is a tricky subject which changes a lot depending on the device manufacturer and the Android version.

biodranik commented on android/src/com/mapswithme/maps/MwmActivity.java:
```diff
@@ -1426,9 +1453,12 @@ private void showMainMenu(boolean show)
-    int totalHeight = calcFloatingViewsOffset();
-    adjustCompassAndTraffic(!show ? UiUtils.getStatusBarHeight(getApplicationContext())
-                                  : totalHeight);
+    int offset;
```
```suggestion
    int offset = 0;
```
> Local variables are slightly different; the compiler never assigns a default value to an uninitialized local variable. If you cannot initialize your local variable where it is declared, make sure to assign it a value before you attempt to use it. Accessing an uninitialized local variable will result in a compile-time error.

Source: https://docs.oracle.com/javase/tutorial/java/nutsandbolts/datatypes.html

biodranik commented on android/src/com/mapswithme/maps/MwmActivity.java:
```diff
@@ -1426,9 +1453,12 @@ private void showMainMenu(boolean show)
-                                  : totalHeight);
+    int offset;
+    if (!show || mRoutingPlanInplaceController == null
+        || (offset = mRoutingPlanInplaceController.calcHeight()) == 0)
```
It looks strange to assign offset here and then reassign it immediately in the next line.

vng replied:
If zero offset is not good (like in previous calcFloatingViewsOffset() function), this logic with always call adjustCompassAndTraffic looks suspicious

arnaudvergnet replied:
I'll rework this logic for something simpler. I simply copy-pasted the old logic without trying to refactor it in fear of breaking something.

biodranik commented on android/src/com/mapswithme/maps/routing/NavigationController.java:
```diff
@@ -88,11 +88,20 @@ public NavigationController(AppCompatActivity activity, @NonNull MapButtonsContr
-      view.getLayoutParams().width = mFrame.findViewById(R.id.nav_bottom_sheet).getWidth();
-      return windowInsets;
-    });
+    View navigationBarBackground = mFrame.findViewById(R.id.nav_bottom_sheet_nav_bar);
```
final

biodranik commented on android/src/com/mapswithme/maps/widget/menu/NavMenu.java:
```diff
@@ -137,6 +139,11 @@ private void toggleNavMenu()
     int headerHeight = mHeaderFrame.getHeight();
+    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M)
+    {
+      headerHeight += UiUtils.getLegacyNavigationBarHeight(mHeaderFrame);
```
Commenting on why it is necessary would be helpful.

biodranik commented on android/src/com/mapswithme/util/UiUtils.java:
```diff
@@ -363,72 +357,100 @@ public static boolean isViewTouched(@NonNull MotionEvent event, @NonNull View vi
-    int resourceId = res.getIdentifier("status_bar_height", "dimen", "android");
-    if (resourceId > 0)
-      result = res.getDimensionPixelSize(resourceId);
+    Window window = activity.getWindow();
```
final here and below.

biodranik commented on android/src/com/mapswithme/util/UiUtils.java:
```diff
@@ -363,72 +357,100 @@ public static boolean isViewTouched(@NonNull MotionEvent event, @NonNull View vi
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M)
+    {
+      WindowInsetsControllerCompat wic = WindowCompat.getInsetsController(window, decorView);
+      if (wic != null)
```
Why it can be null here? Is it the best strategy to do nothing if it's null?

arnaudvergnet replied:
`getInsetsController` can return null if the provided view is not attached to a window or has no decor view in its tree. But as I pass the window's decor view this should not happen. But then the function hints the compiler that the value could be null so not checking will raise a warning.

biodranik replied:
In this case, it is way better to explicitly disable the warning and add a comment, that it should not be null here. And crash if it's null. And then fix it properly.

biodranik commented on android/src/com/mapswithme/util/UiUtils.java:
```diff
@@ -363,72 +357,100 @@ public static boolean isViewTouched(@NonNull MotionEvent event, @NonNull View vi
-    int statusBarHeight = getStatusBarHeight(view.getContext());
-    extendViewPaddingTop(view, statusBarHeight);
+    final Resources resources = view.getResources();
+    int resourceId = resources.getIdentifier("status_bar_height", "dimen", "android");
```
final

biodranik commented on android/src/com/mapswithme/util/UiUtils.java:
```diff
@@ -363,72 +357,100 @@ public static boolean isViewTouched(@NonNull MotionEvent event, @NonNull View vi
-    view.setLayoutParams(lp);
+    final Resources resources = view.getResources();
+    final int orientation = resources.getConfiguration().orientation;
+    int resourceId = resources.getIdentifier("navigation_bar_height", "dimen", "android");
```
final

biodranik reviewed (changes requested):
Good, almost there. I will try to test it on my Android 5 and other devices.
@rtsisyk can you please help with testing too?

biodranik commented:
1. Android 5: The status bar is not transparent for the Light theme (not a blocker, but it would be great to fix it).
2. Android 5: There is an unnecessary large empty bar at the bottom in the nav mode. Also somehow I reproduced all top panels shifted up in a way that the top bar is not visible, and only the bottom part of the direction sign (on the top left) is visible.
3. Android 8 and 9: In the routing and nav mode the status bar is drawn incorrectly, there should be an inset for it.
4. Android 9 (maybe Android 8 too): software buttons overlap the navigation panel.

vng commented on android/src/com/mapswithme/maps/MapFragment.java:
```diff
@@ -90,8 +90,6 @@ private void setupWidgets(int width, int height)
                       ANCHOR_LEFT_TOP);
 
-    mCurrentCompassOffsetX = 0;
-    mCurrentCompassOffsetY = UiUtils.getCompassYOffset(requireContext());
```
In this case I don't understand the logic.
- mCurrentCompassOffsetX and mCurrentCompassOffsetY are assigned in setupCompass only
- setupCompass also reads this variables first

mCurrentCompassOffsetX,  mCurrentCompassOffsetY will be uninitialized?

vng replied:
But we should ensure that this call was happened before, somehow ..

arnaudvergnet commented:
> 1. Android 5: The status bar is not transparent for the Light theme (not a blocker, but it would be great to fix it).

This is intended behavior because it is not possible to make the status bar content black on Android 5. I found it better to show a background to keep the status bar readable. And one is needed anyway for the navigation mode because the header background for the next street is white.

> 2. Android 5: There is an unnecessary large empty bar at the bottom in the nav mode. Also somehow I reproduced all top panels shifted up in a way that the top bar is not visible, and only the bottom part of the direction sign (on the top left) is visible.
> 3. Android 8 and 9: In the routing and nav mode the status bar is drawn incorrectly, there should be an inset for it.
> 4. Android 9 (maybe Android 8 too): software buttons overlap the navigation panel.

I cannot reproduce those issues, can you please attach screenshots?

biodranik commented:
There is also one more bug on Android 9:

![image](https://user-images.githubusercontent.com/170263/193779533-b965919c-8f10-4335-a25c-7f8e43557d54.png)

And here are previous bugs:

![image](https://user-images.githubusercontent.com/170263/193779598-3c55ecac-f840-468d-83dc-b2a1f56f0975.png)

![image](https://user-images.githubusercontent.com/170263/193779399-e88b02a4-242e-421f-bee8-5e564a9be318.png)

arnaudvergnet commented:
Thanks I'll see what I can do about this. What model are those devices? System insets behavior can change depending on the vendor so it is a bit tricky.

arnaudvergnet commented:
I managed to reproduce the issues except the Android 5 one.

arnaudvergnet commented:
@biodranik I fixed all the issues! I first tested on Android 11 and 5 devices, but the behavior with `setOnApplyWindowInsetsListener` and `fitsSystemWindows` changed in the Android 11 version. I thought at first the change appeared in API level 23 so I made special cases for those versions. Those checks are now removed for a cleaner approach.

Prior to Android 11, calling `setOnApplyWindowInsetsListener` or setting `fitsSystemWindows` on a view would consume the event and prevent any children from using `setOnApplyWindowInsetsListener`. I refactored the code to call `setOnApplyWindowInsetsListener` only on the necessary elements and thus preventing a large parent from consuming the event for its children.

biodranik commented:
@arnaudvergnet can you please rebase it? Or I can help if you want.

biodranik commented on android/src/com/mapswithme/maps/MwmActivity.java:
```diff
@@ -424,33 +425,34 @@ protected void onSafeCreate(@Nullable Bundle savedInstanceState)
+    UiUtils.setLightStatusBar(this, !(
+        RoutingController.get().isPlanning()
+        || Framework.nativeIsInChoosePositionMode()
+        || ThemeUtils.isNightTheme(this)
```
nit: for performance reasons, it's better to use the more frequent comparison first in OR clauses.

biodranik commented on android/src/com/mapswithme/maps/MwmActivity.java:
```diff
@@ -1551,20 +1551,23 @@ public void onNavigationStarted()
   {
     closeFloatingToolbarsAndPanels(true);
     initNavigationButtons(MapButtonsController.LayoutMode.regular);
+    refreshLightStatusBar();
```
nit: Is it the best way to refresh? Can copy-paste be avoided in a more beautiful way?

Btw, why it should be refreshed every time here, in every callback?

arnaudvergnet replied:
The best way would be to create different fragments for planning/navigation modes. We could then setup the status bar in those fragments instead of relying on callbacks.

As it is not fragments, we need to make sure the status bar color is correctly updated when planning/navigation starts/ends (black text for navigation and white for planning).

biodranik commented on android/src/com/mapswithme/maps/routing/NavigationController.java:
```diff
@@ -112,11 +104,20 @@ public NavigationController(AppCompatActivity activity, @NonNull MapButtonsContr
 
-    UiUtils.extendViewWithStatusBar(mStreetFrame);
-    UiUtils.extendViewMarginWithStatusBar(turnFrame);
+    // Show a blank view below the navbar to hide the menu content
```
Sometimes, when Android 5 (and others) switch from a light to a dark theme on activating navigation mode, I see some white backgrounds at the top and bottom of the screen before correctly themed and drawn elements appear. Is it caused by this hack? Why exactly is it needed? Can it be avoided?

arnaudvergnet replied:
I can't reproduce this issue but as the hack adds a bar tinted with the background color below the navigation bar, if the theme takes time to switch when going into navigation mode that bar will appear white.

This hack is needed to hide the navigation bottom sheet content when below the navigation bar. As the navigation bar is set to translucent, we cannot control its color. Adding a view behind it was the simplest way I found to make it temporarily opaque.

biodranik reviewed (approved):
Great! Checked, it works now. I have a few more questions.

biodranik reviewed (approved):
Ok, let's move on with it, test on real devices, and implement all improvements/refactorings later.

arnaudvergnet commented:
Rebased on latest master.

arnaudvergnet commented:
Should we wait to fix https://github.com/organicmaps/organicmaps/issues/3631 as well?

biodranik commented:
#3631 can be done in a separate PR.
